tests/integration/common/block_utils.rs

```diff
@@ -167,13 +167,23 @@ pub fn os_hints(
     transactions: Vec<InternalTransaction>,
     tx_execution_infos: Vec<TransactionExecutionInfo>,
 ) -> (StarknetOsInput, ExecutionHelperWrapper) {
-    let mut contracts: HashMap<Felt252, ContractState> = blockifier_state
-        .state
-        .address_to_class_hash
-        .keys()
+    let state_diff = blockifier_state.to_state_diff();
+    let deployed_addresses = state_diff.address_to_class_hash.keys().cloned().collect::<HashSet<_>>();
+
+    let initial_addresses = blockifier_state.state.address_to_class_hash.keys().cloned().collect::<HashSet<_>>();
+
+    let addresses = initial_addresses.union(&deployed_addresses);
+
+    let mut contracts: HashMap<Felt252, ContractState> = addresses
+        .into_iter()
         .map(|address| {
+            let contract_hash = if deployed_addresses.contains(address) {
+                Felt252::ZERO
+            } else {
+                to_felt252(&blockifier_state.get_class_hash_at(*address).unwrap().0)
+            };
             let contract_state = ContractState {
-                contract_hash: to_felt252(&blockifier_state.state.address_to_class_hash.get(address).unwrap().0),
+                contract_hash,
```

Comment:
A generic comment over this block of code: I'd prefer if it were isolated in a function. It's not super clear what it does, isolating it + adding doc would help.

Reply:
you mean contract_hash calculation?

Reply:
The whole 20-ish lines of code. But anyway, I'll probably wrap this differently with the coming PR so it's not a problem.